### DISCUSSION TITLE
Runtime Exception at UnixLoader.LoadLibrary() after setting AvaloniaNativeLibraryPath variable

### DISCUSSION THREAD

**Discussion by Sathiyaraman-M** [AvaloniaUI/Avalonia#14258] [Q&A]
## Describe the bug
I am doing an Avalonia macOS project and I have to write platform-specific code, which requires compiling the provided Avalonia.Native.OSX Xcode project and linking the generated dylib with the Avalonia app. But I am getting runtime exception at startup, of `System.Exception` type. From stacktrace, I find the exception is thrown from `UnixLoader.LoadLibrary()` method, after failing to load the library specified at `AvaloniaNativeLibraryPath` variable.

## To Reproduce

1. Compile the `Avalonia.Native.OSX.xcodeproj` to generate the `libAvalonia.Native.OSX.dylib`
2. Place the `libAvalonia.Native.OSX.dylib` at the macOS project root directory.
3. Ensure the dylib is copied to the output directory always, in the Project properties.
4. Set `AvaloniaNativeLibraryPath` variable to `libAvalonia.Native.OSX.dylib`.
5. Run the project.

## Expected behavior
The application should start normally without any problems, given that I am using the dylib generated from the unmodified Objective-C code, provided in the Avalonia repo.

## Screenshots
![image](https://github.com/AvaloniaUI/Avalonia/assets/45333880/d1bfdb30-57ad-468d-ad2c-d4a93c6f7e90)

## Environment

- OS: macOS Sonoma
- Avalonia-Version: 11.0.4

## Additional context
Though I am new to Avalonia, I guess that I might be using incorrect path or placed the dylib in incorrect directory. But I also suspect if that might be due to any bug, as both the macOS executable and the dylib file are present in same directory after build, but fails to load properly at runtime.

**Comment by maxkatz6** [comment]
We usually use this property with an absolute path for testing, so not sure if it is supposed to work like that. You can check where your file is located after publishing in the app folder. 

But more importantly, you shouldn't get any errors from UnixLoader, if you use "net6.0" target or newer. UnixLoader should be compiled only for netstandard builds.

**Comment by maxkatz6** [comment]
Double checked. AvaloniaNativeLibraryPath needs an absolute path.
So you need to adjust it before passing to the options, probably combining it with your executable folder path.

**Comment by Sathiyaraman-M** [comment]
@maxkatz6 Still not working with absolute path

![AvaloniaNativeLibraryPathNotWorking2](https://github.com/AvaloniaUI/Avalonia/assets/45333880/63800acb-d4d7-4cc2-b2d7-2d84ebbfdf95)

I tried using the path `/Users/_macos_user_/Developer/RiderProjects/Project42.Avalonia/src/Rmkec.App.MacOS/libAvalonia.Native.OSX.dylib` with no success but providing `null` successfully fallback to default dylib.

Would be helpful if there is any sample Avalonia macOS project that uses a compiled dylib linked using `AvaloniaNativeLibraryPath`

**Comment by kekekeks** [comment]
It seems that system dynamic library loader is refusing to load your dylib. Check if can load it manually via NativeLibraryEx or P/Invoking dlopen.

**Comment by Sathiyaraman-M** [comment]
@kekekeks Actually I was able to try running the application in an Intel-based MacBook running Ventura and it worked like a charm. But I couldn't achieve the same on a M1-Macbook Pro, running Sonoma.
